### ISSUE TITLE
Using data parallel

### ISSUE THREAD

**Issue by jiayu2B** [pyannote/pyannote-audio#210]
I tried to use data parallel to process the data. However, the training speed haven't go faster.
I think the reason is that the LSTM and Linear layer is not suitable for data parallel. 
What do you think about this issue?

**Comment by hbredin** [comment]
Can you please explain what you mean by "tried to use data parallel"? 
What did you do exactly?

**Comment by jiayu2B** [comment]
I've rewrite pyannote/audio/labeling/tasks/base.py, in the batch_loss()
Beside, I modified some part in models.py, only add nn.ModuleList to all the lists of models
>     def batch_loss(self, batch):
>         """Compute loss for current `batch`
> 
>         Parameters
>         ----------
>         batch : `dict`
>             ['X'] (`numpy.ndarray`)
>             ['y'] (`numpy.ndarray`)
>             ['mask'] (`numpy.ndarray`, optional)
> 
>         Returns
>         -------
>         batch_loss : `dict`
>             ['loss'] (`torch.Tensor`) : Loss
>         """
> 
>         # forward pass
>         X = torch.tensor(batch['X'],
>                          dtype=torch.float32,
>                          device=self.device_)
>         model_ = torch.nn.DataParallel(self.model_, device_ids=[1])
>         model_=model_.to(self.device_)
>         fX = model_(X)
> 
>

**Comment by hbredin** [comment]
I guess calling `model_ = torch.nn.DataParallel(self.model_, device_ids=[1])` for every batch is probably not optimal. Maybe it can be called just once when training starts? 

I have never used `DataParallel` so, if you do manage to speed things up with this, let me know: I'd love to add this to `pyannote.audio`.

**Comment by hbredin** [comment]
Also, it seems that you are only asking for one device in `DataParallel`: `device_ids=[1]`.  
I believe this means that you are trying to parallelize data on just one GPU so this will have no effect, right?

**Comment by jiayu2B** [comment]
Well, I used device_ids=[0,1], and found the effect is not good, so I changed it into [1].

**Comment by hbredin** [comment]
Closing stalled issue. Reopen if needed.
